programs/drift/src/state/lp_pool.rs

```diff
@@ -0,0 +1,1462 @@
+    /// precision: PERCENTAGE_PRECISION
+    pub swap_fee_min: i64,
+    /// max fee charged on swaps to/from this constituent
+    /// precision: PERCENTAGE_PRECISION
+    pub swap_fee_max: i64,
+
+    /// ata token balance in token precision
+    pub token_balance: u64,
+
+    pub last_oracle_price: i64,
+    pub last_oracle_slot: u64,
+
+    pub oracle_staleness_threshold: u64,
+
+    pub flash_loan_initial_token_amount: u64,
+    /// Every swap to/from this constituent has a monotonically increasing id. This is the next id to use
+    pub next_swap_id: u64,
+
+    /// percentable of derivatve weight to go to this specific derivative PERCENTAGE_PRECISION. Zero if no derivative weight
+    pub derivative_weight: u64,
+
+    pub volatility: u64, // volatility in PERCENTAGE_PRECISION 1=1%
+
+    pub constituent_derivative_index: i16, // -1 if a parent index
+
+    pub spot_market_index: u16,
+    pub constituent_index: u16,
+
+    pub decimals: u8,
+    pub bump: u8,
+
+    // Fee params
+    pub gamma_inventory: u8,
+    pub gamma_execution: u8,
+    pub xi: u8,
+    pub _padding: [u8; 5],
+}
+
+impl Size for Constituent {
+    const SIZE: usize = 288;
+}
+
+impl Constituent {
+    /// Returns the full balance of the Constituent, the total of the amount in Constituent's token
+    /// account and in Drift Borrow-Lend.
+    pub fn get_full_balance(&self, spot_market: &SpotMarket) -> DriftResult<i128> {
+        match self.spot_balance.balance_type() {
+            SpotBalanceType::Deposit => self.token_balance.cast::<i128>()?.safe_add(
+                self.spot_balance
+                    .get_token_amount(spot_market)?
+                    .cast::<i128>()?,
+            ),
+            SpotBalanceType::Borrow => self.token_balance.cast::<i128>()?.safe_sub(
+                self.spot_balance
+                    .get_token_amount(spot_market)?
+                    .cast::<i128>()?,
+            ),
+        }
+    }
+
+    pub fn record_swap_fees(&mut self, amount: i128) -> DriftResult {
+        self.total_swap_fees = self.total_swap_fees.safe_add(amount)?;
+        Ok(())
+    }
+
+    /// Current weight of this constituent = price * token_balance / lp_pool_aum
+    /// Note: lp_pool_aum is from LPPool.last_aum, which is a lagged value updated via crank
+    pub fn get_weight(
+        &self,
+        price: i64,
+        spot_market: &SpotMarket,
+        token_amount_delta: i128,
+        lp_pool_aum: u128,
+    ) -> DriftResult<i64> {
+        if lp_pool_aum == 0 {
+            return Ok(0);
+        }
+        let value_usd = self.get_notional(price, spot_market, token_amount_delta, true)?;
+
+        value_usd
+            .safe_mul(PERCENTAGE_PRECISION_I64.cast::<i128>()?)?
+            .safe_div(lp_pool_aum.cast::<i128>()?)?
+            .cast::<i64>()
+    }
+
+    pub fn get_notional(
+        &self,
+        price: i64,
+        spot_market: &SpotMarket,
+        token_amount: i128,
+        is_delta: bool,
```

Comment:
this fn sig is weird, i see:
1) passing in 0 and true
2) passing in a token amount and false

would it make more sense to do Option<i128>

Reply:
i think it'd be better to just write another function to handle the delta case to minimize branching. looks like wherever this is called it's already hardcoding true/false

Reply:
agreed, makes it less confusing. Probably just have two separate fns